rfd/0207-vnet-ssh.md

```diff
@@ -0,0 +1,392 @@
+   1. VNet will determine which clusters the proxy address or custom DNS zone
+      matches. If none, the DNS request is forwarded upstream.
+   1. VNet will apply any matching proxy templates found in the user's
+      `TELEPORT_HOME`.
+   1. VNet will query each matching cluster to see if any SSH servers have a
+      matching SSH server using the `ResolveSSHTarget` rpc.
+   1. If a match is found a free IP will be assigned to that SSH server and it
+      will be returned in an authoritative DNS answer.
+   1. If no matching server is found VNet will respond with `NXDOMAIN`.
+
+#### Connection forwarding
+
+Here is a sequence diagram for an example VNet SSH session with per-session MFA:
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Client as SSH Client
+    participant DNS_Resolver as DNS Resolver
+    participant VNet
+    participant Teleport as teleport.example.com
+    participant SSH_Server as SSH Server
+
+    User->>Client: ssh host.ssh.teleport.example.com
+    Client->>DNS_Resolver: Ask for A record for host.ssh.teleport.example.com
+    DNS_Resolver->>DNS_Resolver: Check own cache
+    DNS_Resolver->>VNet: Ask for A record for host.ssh.teleport.example.com
+    VNet->>VNet: Check own cache
+    VNet->>VNet: Apply proxy templates
+    VNet->>Teleport: Ask if SSH server "host" exists
+    Teleport->>VNet: Reply "host" exists
+    VNet->>VNet: Assign IP 100.64.0.2 to host.ssh.teleport.example.com
+    VNet->>DNS_Resolver: Return A record with IP 100.64.0.2
+    DNS_Resolver->>Client: Return A record with IP 100.64.0.2
+    Client<<->>VNet: Terminated SSH connection to 100.64.0.2
+    VNet->>Teleport: Request per-session MFA certificate
+    Teleport->>VNet: Send MFA challenge
+    VNet->>User: Prompt user to tap MFA
+    VNet->>Teleport: Send MFA response
+    Teleport->>VNet: Send MFA-verified SSH cert
+    VNet<<->>Teleport: Terminated SSH connection to "host"
+    Teleport<<->>SSH_Server: Tunnelled connection to "host"
+```
+
+### Security
+
+The point of VNet is to allow applications running on the local host to connect
+to Teleport resources transparently, without needing to configure a Teleport
+user certificate.
+It has a similar security model to a VPN, once the user securely logs in to
+Teleport on their host, all applications on the host can connect to their
+Teleport resources.
+Where Teleport improves upon the VPN security model here is with per-session
+MFA.
+When per-session MFA is enabled for SSH access in the cluster, the user will
+have to tap their MFA device every time they start a new SSH session to a
+server.
+
+#### Authentication
+
+VNet will effectively run an in-process SSH server to terminate incoming SSH
+connections and forward them to the target host.
+This SSH server will *not* require any client authentication, as mentioned above
+the point of VNet is to allow transparent access to Teleport resources by any
+application running on the local host where the user has logged in.
```

Comment:
Is this really something we're ok with? Is there some system-level hackery we can employ to confirm that the correct user on the machine is the one opening sockets? Could we add/replace some ephemeral key to the user's SSH agent or maybe trust the default public keys in ~/.ssh/ at least?

Reply:
in the latest revision, now that we are setting options in the SSH config anyway, we'll have the client use a specific key that VNet trusts